### PR TITLE
Feature queue

### PR DESCRIPTION
**Pull request by sccolbert**
Should be g2g

### CODE REVIEW

File: src/collections/queue.ts [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+   * #### Complexity
+   * Constant.
+   *
+   * #### Undefined Behavior
+   * Calling `popFront()` on an empty range is undefined.
+   */
+  popFront(): T {
+    assert(!this.isEmpty(), 'QueueRange#popFront(): Range is empty');
+    let node = this._front;
+    this._front = node.next;
```
Comment (blink1073) [inline src/collections/queue.ts]: Clear `node.next` after this?

Comment (blink1073) [inline src/collections/queue.ts]: N/m, I see the distinction that this is the range implementation, it should not destroy the node itself.

**Comment by blink1073** [comment]
LGTM.

**Comment by jasongrout** [comment]
I'm curious where you are planning on using this?

**Comment by sccolbert** [comment]
It's used by the messaging module.

**Comment by jasongrout** [comment]
Ah, I was thinking this was something new you were writing. I didn't realize this was part of the big squeeze.

**Pull request merged**